src/crate/client/cursor.py

```diff
@@ -49,9 +50,47 @@ def execute(self, sql, parameters=None, bulk_parameters=None):
 
         self._result = self.connection.client.sql(sql, parameters,
                                                   bulk_parameters)
-        if "rows" in self._result:
+
+        if "rows" not in self._result:
+            return
+
+        if "col_types" in self._result:
+            self.rows = iter(self._transform_result_types())
+
+        else:
             self.rows = iter(self._result["rows"])
 
+    def _transform_result_types(self):
+        """
+        Generate row items with column values converted to their corresponding
+        native Python types, based on information from `col_types`.
+
+        Currently, only converting to native `datetime` objects is implemented.
+        """
+        datetime_column_types = [11, 15]
+        datetime_columns_mask = [
+            True if col_type in datetime_column_types else False
+            for col_type in self._result["col_types"]
+        ]
+        for row in self._result["rows"]:
+            yield list(self._transform_datetime_columns(row, iter(datetime_columns_mask)))
+
+    @staticmethod
+    def _transform_datetime_columns(row, column_flags):
+        """
+        Convert all designated columns to native Python `datetime` objects.
+        """
+        for value in row:
+            try:
+                flag = next(column_flags)
+            except StopIteration:
+                break
+
+            if flag and value is not None:
+                value = datetime.fromtimestamp(value / 1e3)
```

Comment:
I think using `utcfromtimestamp` is the less error prone option here.

There are two scenarios:

The user stores UTC on the server:

- `fromtimestamp` would be wrong, or at least confusing as it would convert to localtime but without attaching a timezone info.
- `utcfromtimestamp` would be correct

The user stores local timestamps on the server. Unless they also store the timezone in a separate field this is error prone but

- `utcfromtimestamp` would at least preserve the value used on the server
- `fromtimestamp` would be wrong, because it would apply _another_ offset.


It's important that we don't attach a `tzinfo` given that we don't have one.

Reply:
Another option could be to to let users  configure the timezone on the client (with default UTC) and always create timezone aware objects.

The default could turn out to be wrong, but at least the user has an easy option to fix it, and it's easier to spot a incorrectly attached tzone then it is to figure out that a conversion happened and why/where some conversion happend.
It's also way more convenient/less error prone to continue working with tz aware datetimes.